### DISCUSSION TITLE
Why does `next build` firstly create an optimized production build, THEN process the linting and checking validity of types step?

### DISCUSSION THREAD

**Discussion by goustaff** [vercel/next.js#56155] [Help] [answered]
### Summary

I don't understand why _Next.js_ compiles first, THEN checks the validity of the types and lint.

Wouldn't it make more sense to do the opposite in order to invalidate the build as quickly as possible if the types are incorrect or if the linter raises an error?

### Example

> $ next build
> ✓ Creating an optimized production build    
> ✓ Compiled successfully
>   Linting and checking validity of types  ..Failed to compile.
> 
> ./foo/bar.ts:30:14
> Type error: ...

### Why not?

> $ next build
>   Linting and checking validity of types  ..Failed to compile.
> 
> ./foo/bar.ts:30:14
> Type error: ...

**Comment by icyJoseph** [comment, marked as answer]
From:

- https://github.com/vercel/next.js/blob/f95f8c52d550b52e5daa09706fd57e1c9a791029/packages/next/src/build/index.ts#L462

> // For app directory, we run type checking after build. That's because
> // we dynamically generate types for each layout and page in the app
> // directory.

And the type checking options includes the linting and such:

```ts
 const typeCheckingOptions: Parameters<typeof startTypeChecking>[0] = {
        dir,
        appDir,
        pagesDir,
        runLint,
        shouldLint,
        ignoreESLint,
        telemetry,
        nextBuildSpan,
        config,
        cacheDir,
      }
```

**Reply by goustaff** [reply]
Oh, okay. Thank you!

**Reply by icyJoseph** [reply]
No worries mate, Today I Learned something new thanks to you :)
